vote/src/vote_state_view/field_frames.rs

```diff
@@ -0,0 +1,317 @@
+    pub(super) fn read(cursor: &mut std::io::Cursor<&[u8]>) -> Result<Self> {
+        let len = solana_serialize_utils::cursor::read_u64(cursor)
+            .map_err(|_err| VoteStateViewError::ParseError)? as usize;
+        let len = u8::try_from(len).map_err(|_| VoteStateViewError::ParseError)?;
+        let frame = Self { len };
+        cursor.consume(frame.total_item_size());
+        Ok(frame)
+    }
+}
+
+impl ListFrame<'_> for EpochCreditsListFrame {
+    type Item = EpochCreditsItem;
+
+    fn len(&self) -> usize {
+        self.len as usize
+    }
+}
+
+impl EpochCreditsItem {
+    #[inline]
+    pub fn epoch(&self) -> u64 {
+        u64::from_le_bytes(self.epoch)
+    }
+    #[inline]
+    pub fn credits(&self) -> u64 {
+        u64::from_le_bytes(self.credits)
+    }
+    #[inline]
+    pub fn prev_credits(&self) -> u64 {
+        u64::from_le_bytes(self.prev_credits)
+    }
+}
+
+impl From<&EpochCreditsItem> for (Epoch, u64, u64) {
+    fn from(item: &EpochCreditsItem) -> Self {
+        (item.epoch(), item.credits(), item.prev_credits())
+    }
+}
+
+pub(super) struct RootSlotView<'a> {
+    frame: RootSlotFrame,
+    buffer: &'a [u8],
+}
+
+impl<'a> RootSlotView<'a> {
+    pub(super) fn new(frame: RootSlotFrame, buffer: &'a [u8]) -> Self {
+        Self { frame, buffer }
+    }
+}
+
+impl RootSlotView<'_> {
+    pub(super) fn root_slot(&self) -> Option<Slot> {
+        if !self.frame.has_root_slot {
+            None
+        } else {
+            let root_slot = {
+                let mut cursor = std::io::Cursor::new(self.buffer);
+                cursor.consume(1);
+                solana_serialize_utils::cursor::read_u64(&mut cursor).unwrap()
+            };
+            Some(root_slot)
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+#[cfg_attr(feature = "frozen-abi", derive(AbiExample))]
+pub(super) struct RootSlotFrame {
+    has_root_slot: bool,
+}
+
+impl RootSlotFrame {
+    pub(super) fn total_size(&self) -> usize {
+        1 + self.size()
+    }
+
+    pub(super) fn size(&self) -> usize {
+        if self.has_root_slot {
+            core::mem::size_of::<Slot>()
+        } else {
+            0
+        }
+    }
+
+    pub(super) fn read(cursor: &mut std::io::Cursor<&[u8]>) -> Result<Self> {
+        let has_root_slot = solana_serialize_utils::cursor::read_bool(cursor)
+            .map_err(|_err| VoteStateViewError::ParseError)?;
+        let frame = Self { has_root_slot };
+        cursor.consume(frame.size());
+        Ok(frame)
+    }
+}
+
+pub(super) struct PriorVotersFrame;
+
+impl ListFrame<'_> for PriorVotersFrame {
+    type Item = PriorVotersItem;
+
+    fn len(&self) -> usize {
+        const MAX_ITEMS: usize = 32;
+        MAX_ITEMS
+    }
+
+    fn total_size(&self) -> usize {
+        self.total_item_size() +
+            core::mem::size_of::<u64>() /* idx */ +
+            core::mem::size_of::<bool>() /* is_empty */
+    }
+}
+
+#[repr(C)]
+pub(super) struct PriorVotersItem {
```

Comment:
this one. see above safety comment on `read_item`

Reply:
Nice, just noticed this one as well, thanks for calling it out

Reply:
Fixed here: https://github.com/anza-xyz/agave/pull/4834/commits/a3d48ac49ca336a228813c9f7cf9e6712a2faa4a